Security second part

ShastkivRuslan opened the pull request:
Add Spring Security configuration and Role-based access control

- Integrate spring-boot-starter-security and create SecurityConfig with PasswordEncoder and SecurityFilterChain beans
- Implement UserDetails interface in User entity and create UserDetailsService implementation
- Make /api/auth/register endpoint public and restrict other endpoints based on user roles
- Create Role entity, RoleRepository, and establish Many-to-Many relationship between User and Role entities
- Add changeset to create roles in the database and assign roles to users
- Configure Swagger to remain accessible
- Update User entity to include roles, and use `@PreAuthorize` for role-based access control on endpoints

ShastkivRuslan committed:
-added security starter to pom.xml;
-added SecurityConfig class;

ShastkivRuslan committed:
-implemented UserDetails interface in user class;

ShastkivRuslan committed:
-added user password encoding in the registration method;

ShastkivRuslan committed:
-added findByEmail method to UserRepository;

ShastkivRuslan committed:
-refactored imports;
mvn clean package -> done;

ShastkivRuslan committed:
-implemented UserDetailsService in UserDetailsServiceImpl.class;

ShastkivRuslan committed:
-added the Role entity with id and RoleName fields;

ShastkivRuslan committed:
-added Liquibase script to create 'roles' table for Role entity.
-bugfix in Role class;

ShastkivRuslan committed:
-created RoleRepository interface;

ShastkivRuslan committed:
-added relation to User entity with role entity;

ShastkivRuslan committed:
-added soft-delete concept to User entity;

ShastkivRuslan committed:
-added empty line to liquibase script;
-added 3 liquibase scripts;

ShastkivRuslan committed:
-bugfix;

ShastkivRuslan committed:
-added empty line to some liquibase scripts;
-added Liquibase script that inserts users into the database

ShastkivRuslan committed:
-added Liquibase script that inserts users roles into the database

ShastkivRuslan committed:
-added PreAuthorize annotations;

ShastkivRuslan committed:
-bugfix;

Elena-Bruyako commented on pom.xml:
```diff
@@ -84,6 +84,11 @@
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-security</artifactId>
+            <version>3.3.5</version>
```
```suggestion
```

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/controller/BookController.java:
```diff
@@ -62,6 +81,7 @@ public BookDto createBook(@RequestBody @Valid CreateBookRequestDto bookDto) {
     )
+    @PreAuthorize("hasRole('ADMIN')")
     @PutMapping("/{id}")
     public BookDto updateBook(@PathVariable Long id, @RequestBody CreateBookRequestDto requestDto) {
```
```suggestion
    public BookDto updateBook(@PathVariable Long id, @RequestBody @Valid CreateBookRequestDto requestDto) {
```

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/controller/BookController.java:
```diff
@@ -42,17 +44,34 @@ public List<BookDto> getAll(Pageable pageable) {
+    )
+    @PreAuthorize("hasRole('USER')")
+    @GetMapping("/search")
+    public List<BookDto> search(BookSearchParametersDto searchParametersDto, Pageable pageable) {
```
check other Pageable
```suggestion
    public List<BookDto> search(BookSearchParametersDto searchParametersDto, @ParameterObject @PageableDefault Pageable pageable) {
```

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/model/Role.java:
```diff
@@ -0,0 +1,33 @@
+import jakarta.persistence.Table;
+import org.springframework.security.core.GrantedAuthority;
+
+@Entity
```
```suggestion
@Entity
@Getter
@Setter
```

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/model/User.java:
```diff
@@ -31,4 +42,50 @@ public class User {
+            joinColumns = @JoinColumn(name = "user_id"),
+            inverseJoinColumns = @JoinColumn(name = "role_id")
+    )
+    private Set<Role> roles;
```
```suggestion
    private Set<Role> roles = new HashSet<>();
```

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/repository/user/UserRepository.java:
```diff
@@ -1,8 +1,14 @@
 public interface UserRepository extends JpaRepository<User, Long> {
     boolean existsByEmail(String email);
+
+    @Query("SELECT u FROM User u JOIN FETCH u.roles WHERE u.email = :email")
+    Optional<User> findUserWithRolesByEmail(@Param("email") String email);
```
```suggestion
    @EntityGraph(attributePaths = "roles")
    Optional<User> findByEmail(String email);
```

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/security/UserDetailsServiceImpl.java:
```diff
@@ -0,0 +1,22 @@
+
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
+        return userRepository.findUserWithRolesByEmail(email).orElseThrow(
```
```suggestion
        return userRepository.findByEmail(email).orElseThrow(
```

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/security/UserDetailsServiceImpl.java:
```diff
@@ -0,0 +1,22 @@
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
+        return userRepository.findUserWithRolesByEmail(email).orElseThrow(
+                () -> new UsernameNotFoundException("Can`t find user by email ["
```
```suggestion
                () -> new EntityNotFoundException("Can`t find user by email ["
```

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/service/user/UserServiceImpl.java:
```diff
@@ -22,7 +24,9 @@ public UserDto register(UserRegistrationRequestDto registrationRequestDto) {
             );
         }
-        User savedUser = userRepository.save(userMapper.toModel(registrationRequestDto));
+        User userFromRequest = userMapper.toModel(registrationRequestDto);
+        userFromRequest.setPassword(passwordEncoder.encode(registrationRequestDto.password()));
+        User savedUser = userRepository.save(userFromRequest);
```
set USER role to user

```suggestion
        User user = userMapper.toModel(registrationRequestDto);
        user.setPassword(passwordEncoder.encode(registrationRequestDto.password()));
        userRepository.save(user);
```

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/service/user/UserServiceImpl.java:
```diff
@@ -22,7 +24,9 @@ public UserDto register(UserRegistrationRequestDto registrationRequestDto) {
+        User userFromRequest = userMapper.toModel(registrationRequestDto);
+        userFromRequest.setPassword(passwordEncoder.encode(registrationRequestDto.password()));
+        User savedUser = userRepository.save(userFromRequest);
         return userMapper.toDto(savedUser);
```
```suggestion
        return userMapper.toDto(user);
```

Elena-Bruyako commented on src/main/resources/db/changelog/changes/02-create-users-table.yaml:
```diff
@@ -36,4 +36,4 @@ databaseChangeLog:
                   name: shipping_address
-                  type: varchar(255)
\ No newline at end of file
+                  type: varchar(255)
```
add isDeleted column

Elena-Bruyako commented on src/main/resources/db/changelog/changes/06-add-is_deleted-column.yaml:
```diff
@@ -0,0 +1,14 @@
+databaseChangeLog:
+  - changeSet:
+      id: add-is_deleted-column
```
remove

Elena-Bruyako reviewed (changes requested):
see comments

ShastkivRuslan committed:
Applied changes based on mentor review feedback

-removed 06-add-is_deleted-column.yaml migration script;
-removed spring security version from pom.xml;
-added @Valid annotation to updateBook() method;
-Add @ParameterObject and @PageableDefault annotations to required methods for pageable support;
-added @Getter and @Setter annotations
 to Role entity;
 -initialized roles field as HashSet in User entity;
 -refactored findUserWithRolesByEmail query to use @EntityGraph for fetching roles by email;
 -refactored exception from UsernameNotFoundException to EntityNotFoundException for user retrieval by email;
 -refactored user registration flow: simplify User creation and encoding process;
 -added logic to assign default role 'USER' to newly registered users;
 -added isDeleted column to 02-create-users-table.yaml script;

 mvn clean package -> done;
 build -> success;
 manual testing of endpoints -> done.

liliia-ponomarenko commented on src/main/java/ruslan/shastkiv/bookstore/controller/BookController.java:
```diff
@@ -32,8 +35,9 @@ public class BookController {
+    @PreAuthorize("hasRole('USER')")
     @GetMapping
-    public List<BookDto> getAll(Pageable pageable) {
+    public List<BookDto> getAll(@ParameterObject @PageableDefault Pageable pageable) {
```
Better return Page instead List

liliia-ponomarenko commented on src/main/resources/db/changelog/changes/02-create-users-table.yaml:
```diff
@@ -36,4 +36,10 @@ databaseChangeLog:
-                  type: varchar(255)
\ No newline at end of file
+                  type: varchar(255)
+              - column:
+                  name: is_deleted
+                  type: tinyint(1)
+                  defaultValueBoolean: false
+                  constraints:
+                    nullable: false
```
Better create a new changeSet. On a real project, you would get a checksum mismatch error

liliia-ponomarenko reviewed (approved):
Great 🔥 A few recommendations ;)
